### PR TITLE
feature/SOF-3517

### PR DESCRIPTION
**Pull request by mohammadidinani**

### CODE REVIEW

File: endpoints/jobs.py [hunk 1]
```diff
@@ -1,104 +1,110 @@
...
-            jid (str): job ID.
+            id_ (str): job ID.
         """
-        return self.request('DELETE', '/'.join((self.name, jid)), headers=self.headers)
-
-    def update_job(self, jid, kwargs):
+        self.request('POST', '/'.join((self.name, id_, "submit")), headers=self.headers)
+
+    def get_config(self, material_ids, workflow_id, project_id, owner_id, name, compute=None, is_multi_material=False):
+        config = {
```
Comment (timurbazhirov) [inline endpoints/jobs.py]: Docstring

File: endpoints/jobs.py [hunk 2]
```diff
@@ -1,104 +1,110 @@
...
+        }
+
+        if compute: config.update({"compute": compute})
+        if is_multi_material:
+            config.update({"_materials": [{"_id": id} for id in material_ids]})
+        else:
+            config.update({"_material": {"_id": material_ids[0]}})
+        return config
+
+    def get_compute(self, cluster, ppn=1, nodes=1, queue="D", time_limit="01:00:00", notify="abe"):
```
Comment (timurbazhirov) [inline endpoints/jobs.py]: Docstring

File: endpoints/materials.py [hunk 3]
```diff
@@ -1,116 +1,88 @@
...
+            dict
         """
-        headers = dict([('Content-Type', 'application/json')])
-        headers.update(self.headers)
-        return self.request('PATCH', '/'.join((self.name, mid)), data=json.dumps(kwargs), headers=headers)
+        data = {"name": name, "content": content, "format": format, "ownerId": owner_id, "tags": tags}
+        return self.request('POST', "/".join((self.name, "import")), headers=self.headers, data=json.dumps(data))
 
-    def create_material(self, material):
+    def _get_materialsproject_url(self, material_id):
```
Comment (timurbazhirov) [inline endpoints/materials.py]: Any external functionality should be isolated. In utils, perhaps.

File: endpoints/materials.py [hunk 4]
```diff
@@ -1,116 +1,88 @@
...
+        conn = BaseConnection()
+        with conn:
+            for material_id in material_ids:
+                conn.request("GET", self._get_materialsproject_url(material_id), params={"API_KEY": api_key})
+                material = conn.json()["response"][0]
+                tags.extend(material.get("tags", []))
+                materials.append(self.import_from_file(material["material_id"], material["cif"], owner_id, "cif", tags))
+        return materials
+
+    def flatten_material(self, material):
```
Comment (timurbazhirov) [inline endpoints/materials.py]: Docstring

File: endpoints/materials.py [hunk 5]
```diff
@@ -1,116 +1,88 @@
...
-             dict: new material.
+            str
         """
-        headers = dict([('Content-Type', 'application/json')])
-        headers.update(self.headers)
-        return self.request('POST', self.name, data=json.dumps(material), headers=headers)
+        url = ":".join((MATERIALSPROJECT_HOST, str(MATERIALSPROJECT_PORT)))
+        return urlparse.urljoin(url, "/".join(("rest", MATERIALSPROJECT_VERSION, "materials", material_id, "vasp")))
+
+    def import_from_materialsproject(self, api_key, material_ids, owner_id=None, tags=()):
```
Comment (timurbazhirov) [inline endpoints/materials.py]: Same as above + Missing Docstring

File: examples/run-dft-workflow.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+    # wait for jobs to finish and get the final jobs
+    wait_for_jobs_to_finish(job_endpoints, jobs)
+    jobs = job_endpoints.list(query={"_id": {"$in": [j["_id"] for j in jobs]}})
+
+    # form final table
+    keys = ["ID", "NAME", "TAGS", "N-SITES", "LAT-A", "LAT-B", "LAT-C", "LAT-ALPHA", "LAT-BETA", "LAT-GAMMA"]
+    headers = ["-".join(("INI", key)) for key in keys]
+    headers.extend(["-".join(("FIN", key)) for key in keys])
+    headers.extend(["PRESSURE", "DIRECT-GAP", "INDIRECT-GAP"])
```
Comment (timurbazhirov) [inline examples/run-dft-workflow.py]: The length of this file is too long - simplify it, isolate concepts

File: examples/run-dft-workflow.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        band_gap_indirect = raw_property_endpoints.get_indirect_band_gap(job["_id"], unit_flowchart_id)
+
+        # form table
+        data = material_endpoints.flatten_material(initial_structure)
+        data.extend(material_endpoints.flatten_material(final_structure))
+        data.extend([pressure, band_gap_direct, band_gap_indirect])
+
+        rows.append(data)
+
+    print tabulate(rows, headers, tablefmt='grid', stralign='center')
```
Comment (timurbazhirov) [inline examples/run-dft-workflow.py]: We need to get CSV in the end

File: examples/run-dft-workflow.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        # extract final structure
+        unit_flowchart_id = job["workflow"]["subworkflows"][0]["units"][0]["flowchartId"]
+        final_structure = raw_property_endpoints.get_property(job["_id"], unit_flowchart_id, "final_structure")["data"]
+
+        # extract pressure
+        unit_flowchart_id = job["workflow"]["subworkflows"][0]["units"][0]["flowchartId"]
+        pressure = raw_property_endpoints.get_property(job["_id"], unit_flowchart_id, "pressure")["data"]["value"]
+
+        # extract band gaps
+        unit_flowchart_id = job["workflow"]["subworkflows"][1]["units"][1]["flowchartId"]
```
Comment (timurbazhirov) [inline examples/run-dft-workflow.py]: It is difficult to understand how/where the indices are coming from

File: examples/run-ml-workflow.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    predict_material_properties = predicted_properties["data"]["values"][predict_material["exabyteId"]]
+    band_gaps = next((v for v in predict_material_properties if v["name"] == "band_gaps"))
+
+    band_gap_direct = next((v for v in band_gaps["values"] if v["type"] == "direct"), None)["value"]
+    band_gap_indirect = next((v for v in band_gaps["values"] if v["type"] == "indirect"), None)["value"]
+
+    # print results
+    header = ["DIRECT-GAP", "INDIRECT-GAP"]
+    rows = [[band_gap_direct, band_gap_indirect]]
+    print tabulate(rows, header, tablefmt='grid', stralign='center')
```
Comment (timurbazhirov) [inline examples/run-ml-workflow.py]: Same as above:

- too long - split into smaller pieces
- difficult to understand the indices on line 71
- no CSV at the end

File: endpoints/jobs.py (hunk 1 above)
Comment (mohammadidinani) [reply endpoints/jobs.py]: Done.

File: endpoints/jobs.py (hunk 2 above)
Comment (mohammadidinani) [reply endpoints/jobs.py]: Done.

File: endpoints/materials.py (hunk 3 above)
Comment (mohammadidinani) [reply endpoints/materials.py]: Done.

File: endpoints/materials.py (hunk 5 above)
Comment (mohammadidinani) [reply endpoints/materials.py]: Done.

File: endpoints/materials.py (hunk 4 above)
Comment (mohammadidinani) [reply endpoints/materials.py]: Done.

**Pull request merged**